ENH: automatic lag selection for Box-Pierce, Ljung-Box #6645

w31ha0 opened the pull request:
- [x] closes #xxxx
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

With reference to issue [#6645](https://github.com/statsmodels/statsmodels/issues/6645)

Automatic lag selection according to the reference below for Box-Pierce, Ljung-Box. According to the reference at page8, it introduces a penalty term, which switches between AIC and BIC depending on the value of correlation. This penalty term penalises the test statistic and the optimal lag length is one which is able to maximise the test statistic after applying the penalty.

A q value of 2.4 is recommended in the reference for the penalty term according to other research results quoted.

[Reference](https://www.sciencedirect.com/science/article/abs/pii/S0304407609000773?via%3Dihub#!)

<details>


**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in master and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/master -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

pep8speaks commented:
Hello @w31ha0! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`statsmodels/stats/diagnostic.py`](https://github.com/statsmodels/statsmodels/blob/9e64e32da52981af86b90f3a64b5e43a93fba89b/statsmodels/stats/diagnostic.py):

> [Line 554:5](https://github.com/statsmodels/statsmodels/blob/9e64e32da52981af86b90f3a64b5e43a93fba89b/statsmodels/stats/diagnostic.py#L554): [E306](https://duckduckgo.com/?q=pep8%20E306) expected 1 blank line before a nested definition, found 0

bashtage commented:
Looks good overall. Doc string needs to follow the standard format with Parameters and Returns.

bashtage commented:
Should this just be an option for acorr_lungbox? Seems very similar since it just adds a data driven method to select a lag for the test.

w31ha0 commented:
> Should this just be an option for acorr_lungbox? Seems very similar since it just adds a data driven method to select a lag for the test.

Makes sense. I have removed the new method and added a new bool flag (auto_lag) into the existing method.

w31ha0 commented:
@bashtage When does this usually get merged?

bashtage commented:
@w31ha0 Thanks for this contirbution!
